doc/gold.rst

```diff
@@ -280,30 +301,98 @@ introduces a loop which executes the provided block once for each element of the
 value of the expression, with the provided identifier bound to the value of the
 element.
 
-The provided block introduces a new scope.
+The provided block introduces a new scope. If an identifier is provided, then
+the block is named with that identifier. Otherwise, the block is anonymous.
 
 
-``break``
-~~~~~~~~~
+``break`` & ``continue``
+~~~~~~~~~~~~~~~~~~~~~~~~
 
 Usage: ::
 
-  break
+  break [<identifier>]
+  continue [<identifier>]
 
-Terminates the innermost loop currently executing. It is an error to have a
-``break`` statement outside of a loop.
+Terminates a currently-executing block. If no identifier is specified, then the
+innermost block is terminated; otherwise, the named block is terminated along
+with all inner blocks.
 
+In the case of ``break``, the associated loop is also terminated, if one is
+present. It is an error to have a ``break`` outside of a block.
 
-``continue``
-~~~~~~~~~~~~
+In the case of ``continue``, the loop is not terminated. It is an error to have
+a ``continue`` outside of a loop block.
+
+
+``redo``
+~~~~~~~~
 
 Usage: ::
 
-  continue
+  redo [[<identifier>][([<expr>[, ...])]]]
+
+Terminates the currently-executing block, then re-starts it. Rules for the
+identifier are as ``break`` and ``continue``.
+
+If the block in question is a ``fun``, then an argument list may be provided,
+with arguments of the same number and type of the function. The arguments will
+be rebound to the new values. If no argument list is provided, the function
+arguments will retain their current bindings.
 
-Terminates the currently-executing block, but does not terminate the loop,
-instead causing it to move to the next iteration if any remain. It is an error
-to have a ``continue`` statement outside of a loop.
+If the block in question is a ``let`` statement, then an argument list may be
+provided, with arguments of the same number and type as let-bindings introduced.
+The let-bound variables will be rebound with the new values.
+
+All looping constructs may be expressed in terms of ``do``, ``let``, ``if``, and
+``redo``. For example ::
+
+  while <cond> do
+    [...]
+  end
+
+  do if <cond> then
+    [...]
+    redo
+  end end
+
+  for <identifier>: <type> in <start> to <end> do
+    [...]
+  end
+
+  let index: <type> = <start> do: loop
+    if index < <end> then
+      let <identifier>: <type> = index
+      [...]
+      redo loop(index + 1)
+    end
+  end
+
+  for <identifier>: <type> in <expr> do
+    [...]
+  end
+
+  let slice: &[<type>] = <expr> do
+    let index: u16 = mem.start(slice) do: loop
+      if index < mem.end(slice) then
+        let <identifier>: <type> = slice[index]
+        [...]
+        redo loop(index + 1)
+      end
+    end
+  end
+  
+Here is a simple example of a factorial function, expressed using what is
+effectively tail-recursion: ::
+
+  fun factorial(n: u16) -> u16
+    let acc: u16, k: u16 = 1, n do
```

Comment:
i think this is missing a `mut`?

Reply:
I don't think it is, but I might be wrong. All that "immutable" means is that "you can take a pointer to this, and as long as it's still valid, the result of dereferencing that pointer won't change".

In the factorial example, this is actually true; if you take a pointer to, say, `acc`, then you must be taking the pointer inside the let-do-block. However, that pointer is only valid to the end of the scope. The `redo` statement has a subtlety, here -- since it ends the block, it closes the scope, but until the scope is closed, the bindings keep the same value. It then re-starts the block, and re-opens the scope, this time with the new memory values. In other words, calling `redo` necessarily invalidates the pointer.

```
fun foo(n: u16) -> u16
  let k: u16, i: u16 = n, n do
    let j: &u16 = &k
    if k > 1 then
      /* this is fine */
      redo (i - 1, *j - 1)
    else
      return k
    end
  end
end
```

Notice above--the pointer `j` is actually valid until the next statement executed after the `redo` statement, and can be used safely! The memory used by `k` isn't changed until then -- in fact, `k` isn't even rebound until the `redo` statement completes. The fact that we re-use the memory for `k` and `i` should be an implementation detail; by the time we re-use it, you no longer have normal access to it anyway.

I agree that this is a little confusing at first, but I don't want to make the use of `redo` require you to smear gratuitous `mut`s everywhere; in fact, I kind of see it as an alternative to `mut`. It gives you two options for an algorithm:

1. Implement the algorithm iteratively, explicitly marking the memory that's to be re-used. Memory to be re-used can be changed arbitrarily.
2. Implement the algorithm tail-recursively, allowing the compiler to automatically re-use memory based on scope. Memory to be re-used can only be changed per-iteration.

Which option is clearer depends on the algorithm and the person writing it, of course, though I personally feel like option 2 really is the better option.

Reply:
i see. this is starting to explain some of the advantages to your enhanced redo. i'm still struggling with it, but i'm going to let it sit for a while to be sure my aversion isn't "this is different!" but rather "this could be better".